pkg/pullexecutor/pullexecutor.go

```diff
@@ -0,0 +1,147 @@
+	"github.com/warm-metal/csi-driver-image/pkg/pullstatus"
+	"github.com/warm-metal/csi-driver-image/pkg/remoteimage"
+	"github.com/warm-metal/csi-driver-image/pkg/secret"
+	"k8s.io/apimachinery/pkg/util/wait"
+	cri "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+	"k8s.io/klog/v2"
+)
+
+const (
+	pullPollTimeInterval = 100 * time.Millisecond
+	pullPollTimeout      = 2 * time.Minute
+	pullCtxTimeout       = 10 * time.Minute
+)
+
+// PullExecutorOptions are the options passed to the pull executor
+type PullExecutorOptions struct {
+	AsyncPull          bool
+	ImageServiceClient cri.ImageServiceClient
+	SecretStore        secret.Store
+	Mounter            *backend.SnapshotMounter
+}
+
+// PullOptions are the options for a single pull request
+type PullOptions struct {
+	// Context here is only valid for synchronous mounts
+	Context     context.Context
+	NamedRef    docker.Named
+	PullAlways  bool
+	PullSecrets map[string]string
+	Image       string
+}
+
+// PullExecutor executes the pulls
+type PullExecutor struct {
+	asyncPull      bool
+	imageSvcClient cri.ImageServiceClient
+	mutex          *sync.Mutex
+	asyncErrs      map[docker.Named]error
+	secretStore    secret.Store
+	mounter        *backend.SnapshotMounter
+}
+
+// NewPullExecutor initializes a new pull executor object
+func NewPullExecutor(o *PullExecutorOptions) *PullExecutor {
+	return &PullExecutor{
+		asyncPull:      o.AsyncPull,
+		mutex:          &sync.Mutex{},
+		imageSvcClient: o.ImageServiceClient,
+		secretStore:    o.SecretStore,
+	}
+}
+
+// StartPulling starts pulling the image
+func (m *PullExecutor) StartPulling(o *PullOptions) error {
+
+	keyring, err := m.secretStore.GetDockerKeyring(o.Context, o.PullSecrets)
+	if err != nil {
+		return errors.Errorf("unable to fetch keyring: %s", err)
+	}
+
+	if !m.asyncPull {
+		puller := remoteimage.NewPuller(m.imageSvcClient, o.NamedRef, keyring)
+		shouldPull := o.PullAlways || !m.mounter.ImageExists(o.Context, o.NamedRef)
+		if shouldPull {
+			klog.Infof("pull image %q ", o.Image)
+			if err = puller.Pull(o.Context); err != nil {
+				pullstatus.Update(o.NamedRef, pullstatus.Errored)
+				return errors.Errorf("unable to pull image %q: %s", o.Image, err)
+			}
+		}
+		return nil
+	}
+
+	if pullstatus.Get(o.NamedRef) == pullstatus.Pulled ||
+		pullstatus.Get(o.NamedRef) == pullstatus.StillPulling {
+		return nil
+	}
+
+	go func() {
+		if pullstatus.Get(o.NamedRef) == pullstatus.StatusNotFound {
+			m.mutex.Lock()
+			c, cancel := context.WithTimeout(context.Background(), pullCtxTimeout)
+			defer cancel()
+
+			if pullstatus.Get(o.NamedRef) == pullstatus.StillPulling {
+				m.mutex.Unlock()
+				return
+			}
+
+			puller := remoteimage.NewPuller(m.imageSvcClient, o.NamedRef, keyring)
+			shouldPull := o.PullAlways || !m.mounter.ImageExists(o.Context, o.NamedRef)
+			if shouldPull {
+				klog.Infof("pull image %q ", o.Image)
+				pullstatus.Update(o.NamedRef, pullstatus.StillPulling)
+				if err = puller.Pull(c); err != nil {
```

Comment:
So, it seems like we have pushed the deadline of 2m to 10m (controlled via constant), but it's still not having the similar behavior like kubelet which supports - 

* True background based context
* Rate limiting per second
* Max # of inflight pull requests
  
It might result into feature creep if we go all the features that kubelet has, but atleast the last one is something I wish for. We want to control the exposure to the disks with 3) option in async mode.

Reply:
Second point is - How are we controlling image pulling activity on containerd side on warm-metal container restarts ? Will the context that we are passing to containerd gets cancelled ?

Might still want to support graceful parent context handling on container termination or on shutdown.

Reply:
> True background based context

Not sure what `True` background based context means here. 

> Rate limiting per second

I think making the request async is going to ease the exposure to disks. If we don't get enough improvement with it, we can have a follow-up PR to control the number of in-flight requests. 

> Second point is - How are we controlling image pulling activity on containerd side on warm-metal container restart ? Will the context that we are passing to containerd gets cancelled ?

If the context deadline of 10m gets exceeded, we'd get an error but if the image gets pulled, the [context would get cancelled](https://github.com/warm-metal/csi-driver-image/pull/71/files#diff-d02af13e9832014d9034180b6c5d1fbca0ee1c7a7de6f1bb5baa6baf94bef1c2R93-R94). 

> Might still want to support graceful parent context handling on container termination or on shutdown.

Not sure I fully understand this. I guess you are asking what happens if the driver is serving request and the driver pod gets termination signal?

Reply:
> Max # of inflight pull requests

I find the idea interesting but on a second thought I am not sure if this is going to help as much as we think it would. I think the benefit of inflight requests is when containerd is under a lot of load (via kubelet) because there is a lot of pod activity on the node. In such a scenario, the max time we can keep the requests in-flight is 2 minutes (hard limit by kubelet) and I highly doubt the high pod activity is going to be over within 2 minutes in which case, the only real benefit I see here is, we would ease the load on the disks is by adding a short delay between processing of requests and thereby spreading the load over the disk but the benefit here can be argued because all the requests are not going to hit the driver at once (all the pulls won't finish at the same time either), which should give us the randomizing effect we need on disk operations anyway.

Reply:
By True background context, I meant a context without any deadline - `context.Background()` or with really long timeout. 10m is long enough timeout for a image, so it's okay for now.

> If the context deadline of 10m gets exceeded, we'd get an error but if the image gets pulled, the [context would get cancelled](https://github.com/warm-metal/csi-driver-image/pull/71/files#diff-d02af13e9832014d9034180b6c5d1fbca0ee1c7a7de6f1bb5baa6baf94bef1c2R93-R94).

> Not sure I fully understand this. I guess you are asking what happens if the driver is serving request and the driver pod gets termination signal?

Yes, I mean what if the driver gets a termination request and wants to restart. In that case, the inflight grpc requests that being sent to containerd might not get cancelled. To achieve that, you have to build a new context with the parent context or have some cancel mechanism to kill the inflight pull requests on driver restart to support graceful stop. 

> I think making the request async is going to ease the exposure to disks. If we don't get enough improvement with it, we can have a follow-up PR to control the number of in-flight requests.

IMO controlling max # of inflight requests is more important than relaxing the deadline and i'm keen to land on this feature for warm-metal. Currently even on 56 pods, sometimes our disk setup backed by RAID is not able to keep up with all the image pull/unpack operations required by workloads.  It becomes an important scaling factor/lever when we will go for larger nodes with higher pod churn ( =~ 150-200 ) both on kubelet and warm-metal. Kubelet is going to land this feature in 1.27 (ref: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3673-kubelet-parallel-image-pull-limit) and we would limit it to 50 ( or somewhere close).

Reply:
> Yes, I mean what if the driver gets a termination request and wants to restart. In that case, the inflight grpc requests that being sent to containerd might not get cancelled. To achieve that, you have to build a new context with the parent context or have some cancel mechanism to kill the inflight pull requests on driver restart to support graceful stop.

Good point. I don't think we have anything in place right now in the driver to handle cancellation of contexts around pod restart. 

> IMO controlling max # of inflight requests is more important than relaxing the deadline and i'm keen to land on this feature for warm-metal. Currently even on 56 pods, sometimes our disk setup backed by RAID is not able to keep up with all the image pull/unpack operations required by workloads. It becomes an important scaling factor/lever when we will go for larger nodes with higher pod churn ( =~ 150-200 ) both on kubelet and warm-metal. I'm not very hopeful that our current disk setup is going to work with larger nodes.

I am not sure about the importance but feedback noted (we don't have enough metrics/logs to show how much impact warm metal has on CRI; agree with the overall idea of adding rate limiting on warm metal side). I will create a separate ticket for this and try to prioritise it.

Reply:
Created https://github.com/warm-metal/csi-driver-image/issues/77

Resolving the conversation. Feel free to unresolve if needed.